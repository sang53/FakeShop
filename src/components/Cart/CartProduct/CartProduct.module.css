.cartItem {
  height: 500px;
  width: clamp(400px, 50vw, 1000px);
  background-color: var(--cart-bgc);
  border-radius: 10px;
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(350px, 1fr));
  align-items: center;
  justify-items: center;
  box-shadow: var(--product-shadow);

  h1 {
    font-size: 18px;
  }

  > img {
    min-height: 0;
    min-width: 0;
    max-height: 100%;
    border-radius: 15px;
    padding: 1rem;
    background-color: var(--product-img-bgc);
  }

  input {
    margin-inline: 1ch;

    &[type="number"] {
      max-width: 5ch;
    }
  }
}

.info {
  text-align: center;
  display: flex;
  flex-direction: column;
  gap: 1rem;
  align-items: center;
}
