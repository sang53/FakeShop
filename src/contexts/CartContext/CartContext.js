import { createContext } from "react";

export const CartItems = createContext(null);
export const CartReducer = createContext(null);
