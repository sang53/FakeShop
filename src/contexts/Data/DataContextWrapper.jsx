import { useEffect, useState } from "react";
import Helpers from "../../Helper/Helpers";
import { Products } from "./DataContext";
import PropTypes from "prop-types";

export default function DataContextWrapper({ children }) {
  const [data, setData] = useState([]);

  useEffect(() => {
    async function getData() {
      const newData = await Helpers.getProduct();
      if (useData) setData(newData);
    }
    let useData = true;
    getData();

    return () => (useData = false);
  }, []);

  return <Products.Provider value={data}>{children}</Products.Provider>;
}

DataContextWrapper.propTypes = {
  children: PropTypes.element.isRequired,
};
