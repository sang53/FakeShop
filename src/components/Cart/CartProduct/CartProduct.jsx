import classes from "./CartProduct.module.css";
import { use, useRef } from "react";
import { CartReducer } from "../../../contexts/CartContext/CartContext";
import { Products } from "../../../contexts/Data/DataContext";
import PropTypes from "prop-types";
import { Link } from "react-router";
import Helper, { checkout } from "../../../Helper/Helpers";

export default function CartProduct({ id, quantity }) {
  const product = use(Products).find((obj) => obj.id === id);
  const cartDispatch = use(CartReducer);
  const totalRef = useRef(null);

  if (!product) return <h1>Error: Product not found</h1>;

  function handleCart(event) {
    if (event.target.value < 0) return;
    if (event.target.value === "") {
      event.target.value = quantity;
      totalRef.current.value = Helper.getTotal(quantity, product.price);
    } else
      cartDispatch({
        type: "set",
        id,
        quantity: event.target.value,
      });
  }

  return (
    <div className={"gap-1rem pad-1rem " + classes.cartItem}>
      <img src={product.image} alt={product.title} />
      <div className={classes.info}>
        <h1>
          <Link to={`products/${id}`}>{product.title}</Link>
        </h1>
        <div>
          <i>{Helper.padDecimal(product.price)}</i>
        </div>
        <div>
          <label htmlFor={`quantity-${product.id}`} className="bold">
            Quantity:
          </label>
          <input
            id={`quantity-${product.id}`}
            type="number"
            min="0"
            onBlur={handleCart}
            onChange={(e) => {
              if (e.target.value === "")
                totalRef.current.value = Helper.getTotal(
                  quantity,
                  product.price
                );
              else if (e.target.value < 0) totalRef.current.value = "N/A";
              else
                totalRef.current.value = Helper.getTotal(
                  e.target.value,
                  product.price
                );
            }}
            defaultValue={quantity}
          />
        </div>
        <div>
          <label htmlFor="total" className="bold">
            Total:
          </label>
          <input
            id="total"
            className="total"
            type="text"
            readOnly
            ref={totalRef}
            value={Helper.getTotal(quantity, product.price)}
          />
        </div>
        <button
          onClick={(e) => {
            checkout(e.target, { [id]: quantity }, cartDispatch);
          }}
        >
          Checkout
        </button>
        <button onClick={() => cartDispatch({ type: "set", id, quantity: 0 })}>
          Remove
        </button>
      </div>
    </div>
  );
}

CartProduct.propTypes = {
  id: PropTypes.number.isRequired,
  quantity: PropTypes.number.isRequired,
};
