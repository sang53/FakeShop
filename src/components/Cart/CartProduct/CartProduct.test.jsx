import { render, screen } from "@testing-library/react";
import { Products } from "../../../contexts/Data/DataContext";
import { CartReducer } from "../../../contexts/CartContext/CartContext";
import CartProduct from "./CartProduct";
import { describe, expect, test, vi } from "vitest";
import { mockProducts } from "../../../test/utils/mockProducts";
import { MemoryRouter } from "react-router";
import { mockFetch } from "../../../test/utils/utils";

function renderCartProduct(id = 1, quantity = 3) {
  const mockDispatch = vi.fn();
  const renderReturns = render(
    <MemoryRouter initialEntries={["/cart"]}>
      <Products.Provider value={mockProducts}>
        <CartReducer.Provider value={mockDispatch}>
          <CartProduct id={id} quantity={quantity} />
        </CartReducer.Provider>
      </Products.Provider>
    </MemoryRouter>
  );
  return { ...renderReturns, mockDispatch };
}

test("product not found error", () => {
  renderCartProduct(999, 1);
  expect(screen.getByRole("heading")).toHaveTextContent(/error/i);
});

test("cart product elements displayed as expected", () => {
  const { container } = renderCartProduct();
  expect(container).toMatchSnapshot();
});

describe("quantity & total inputs", () => {
  test("initial quantity & total is correct", () => {
    renderCartProduct();
    expect(screen.getByLabelText(/quantity/i)).toHaveValue(3);
    expect(screen.getByLabelText(/total/i)).toHaveValue("$329.85");
  });
  test("change quantity to valid value", async () => {
    const { mockDispatch } = renderCartProduct();
    const quantityInput = screen.getByLabelText(/quantity/i);
    const totalInput = screen.getByLabelText(/total/i);

    // change value without unfocusing
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{5}");
    expect(quantityInput).toHaveValue(5);
    expect(totalInput).toHaveValue("$549.75");
    expect(mockDispatch).not.toHaveBeenCalled();

    // unfocus changed value
    await user.click(totalInput);
    expect(quantityInput).toHaveValue(5);
    expect(totalInput).toHaveValue("$549.75");
    expect(mockDispatch).toHaveBeenCalledOnce();
    expect(mockDispatch.mock.calls[0][0]).toMatchInlineSnapshot(`
      {
        "id": 1,
        "quantity": "5",
        "type": "set",
      }
    `);
  });
  test("change quantity to empty string", async () => {
    const { mockDispatch } = renderCartProduct();
    const quantityInput = screen.getByLabelText(/quantity/i);
    const totalInput = screen.getByLabelText(/total/i);

    // change value without unfocusing
    await user.click(quantityInput);
    await user.keyboard("{Backspace}");
    expect(quantityInput).toHaveValue(null);
    expect(totalInput).toHaveValue("$329.85");
    expect(mockDispatch).not.toHaveBeenCalled();

    // unfocus changed value
    await user.click(totalInput);
    expect(quantityInput).toHaveValue(3);
    expect(totalInput).toHaveValue("$329.85");
    expect(mockDispatch).not.toHaveBeenCalled();
  });
  test("change quantity to 0", async () => {
    const { mockDispatch } = renderCartProduct();
    const quantityInput = screen.getByLabelText(/quantity/i);
    const totalInput = screen.getByLabelText(/total/i);

    // change value without unfocusing
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{0}");
    expect(quantityInput).toHaveValue(0);
    expect(totalInput).toHaveValue("$0.00");
    expect(mockDispatch).not.toHaveBeenCalled();

    // unfocus changed value
    // Note: removed on CartContainer re-render, hence not tested here
    await user.click(totalInput);
    expect(mockDispatch).toHaveBeenCalledOnce();
    expect(mockDispatch.mock.calls[0][0]).toMatchInlineSnapshot(`
      {
        "id": 1,
        "quantity": "0",
        "type": "set",
      }
    `);
  });
  test("change quantity to negative", async () => {
    const { mockDispatch } = renderCartProduct();
    const quantityInput = screen.getByLabelText(/quantity/i);
    const totalInput = screen.getByLabelText(/total/i);

    // change value without unfocusing
    await user.click(quantityInput);
    await user.keyboard("{Backspace}{-}{3}");
    expect(quantityInput).toHaveValue(-3);
    expect(totalInput).toHaveValue("N/A");
    expect(mockDispatch).not.toHaveBeenCalled();

    // unfocus changed value
    await user.click(totalInput);
    expect(quantityInput).toHaveValue(-3);
    expect(totalInput).toHaveValue("N/A");
    expect(mockDispatch).not.toHaveBeenCalled();
  });
});

test("remove button", async () => {
  const { mockDispatch } = renderCartProduct();
  await user.click(screen.getByRole("button", { name: /remove/i }));
  expect(mockDispatch).toHaveBeenCalledOnce();
  expect(mockDispatch.mock.calls[0][0]).toMatchInlineSnapshot(`
    {
      "id": 1,
      "quantity": 0,
      "type": "set",
    }
  `);
});

test("checkout button", async () => {
  const { mockDispatch } = renderCartProduct();
  mockFetch();
  await user.click(screen.getByRole("button", { name: /check/i }));
  expect(mockDispatch).toHaveBeenCalledExactlyOnceWith({
    type: "set",
    id: "1",
    quantity: 0,
  });
});
