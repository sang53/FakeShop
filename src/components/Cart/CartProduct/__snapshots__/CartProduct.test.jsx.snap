// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cart product elements displayed as expected 1`] = `
<div>
  <div
    class="gap-1rem pad-1rem _cartItem_3ded9d"
  >
    <img
      alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
      src="#"
    />
    <div
      class="_info_3ded9d"
    >
      <h1>
        <a
          data-discover="true"
          href="/products/1"
        >
          Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
        </a>
      </h1>
      <div>
        $109.95
      </div>
      <div>
        <label
          for="quantity-1"
        >
          Quantity:
        </label>
        <input
          id="quantity-1"
          min="0"
          type="number"
          value="3"
        />
      </div>
      <div>
        <label
          for="total"
        >
          Total: 
        </label>
        <input
          class="total"
          id="total"
          readonly=""
          type="text"
          value="$329.85"
        />
      </div>
      <button>
        Checkout
      </button>
      <button>
        Remove
      </button>
    </div>
  </div>
</div>
`;
